File: 1458-sort-integers-by-the-number-of-1-bits/1458-sort-integers-by-the-number-of-1-bits.java
class Solution {
    public int[] sortByBits(int[] arr) {
        Map<Integer, List<Integer>> map = new TreeMap<>();
        List<Integer> ans = new ArrayList<>();
        Arrays.sort(arr);

        for(int num: arr){
            int bitCount = Integer.bitCount(num);
            if(!map.containsKey(bitCount)){
                List<Integer> temp = new ArrayList<>();
                temp.add(num);
                map.put(bitCount, temp);
            }else{
                map.get(bitCount).add(num);
            }
        }

        for(int key: map.keySet()){
            ans.addAll(map.get(key));
        }
            
        return ans.stream().mapToInt(Integer::intValue).toArray();
    }
}

File: 1458-sort-integers-by-the-number-of-1-bits/1458-sort-integers-by-the-number-of-1-bits.ts
function sortByBits(arr: number[]): number[] {

    let map: Map<number, number[]> = new Map();
    let ans: number[] = []

    arr.sort((a,b) => a-b);
    for(let num of arr){
        const bit = bitCount(num);
        if(!map.has(bit)){
            map.set(bit, [num]);
        }else{
            map.get(bit).push(num);
        }
    }

    let keys = [...map.keys()].sort((a,b)=> a-b);

    for(const key of keys){
        ans = ans.concat(map.get(key));
    }


    return ans;
};

function bitCount(num:number){
    let bitOne = 0;
    for(let bit of num.toString(2)){
        if(bit == '1') bitOne++;
    }
    return bitOne;
}